tensorflow/core/kernels/reduction_gpu_kernels.cu.h

```diff
@@ -43,6 +43,24 @@ struct SqrtOfReal {
   }
 };
 
+#if TENSORFLOW_USE_ROCM
```

Comment:
Can we just provide the needed overload of Eigen::numext::real such that the existing code works as intended?

Reply:
We probably can, but I'm not sure what's to be gained from that - instead of ROCM-specific templates for SqrtOfReal and Square, we'd have ROCM-specific templates for Eigen::numext::real and Eigen::numext::conj. And they would still have to be guarded by ifdefs since types hipFloatComplex,hipDoubleComplex are not known outside ROCM.

Reply:
The following seems necessary to go the Eigen way:
```
template <>
struct Eigen::internal::real_retval<typename 
  Eigen::internal::global_math_functions_filtering_base<hipFloatComplex>
    ::type> {
  typedef float type;
};

template <>
struct Eigen::internal::real_retval<typename 
  Eigen::internal::global_math_functions_filtering_base<hipDoubleComplex>
    ::type> {
  typedef double type;
};

template <>
__host__ __device__ float Eigen::numext::real(const hipFloatComplex& a) {
  return a.x;
}

template <>
__host__ __device__ double Eigen::numext::real(const hipDoubleComplex& a) {
  return a.x;
}

template <>
__host__ __device__ hipFloatComplex Eigen::numext::conj(const hipFloatComplex& a) {
  return hipFloatComplex(a.x, -a.y);
}
```

Is that how you prefer it?